runtime/domain.c

```diff
@@ -2118,19 +2136,24 @@ void caml_domain_terminate(bool last)
     caml_plat_unlock(&all_domains_lock);
   }
 
-  /* domain_state may be re-used by a fresh domain here (now that we
-   * have done remove_from_stw_domains and released the
-   * all_domains_lock). However, domain_create() won't touch it until
-   * it has claimed the domain_lock, so we hang onto that while we are
-   * tearing down the state. */
+  CAMLassert(last ||
+             (domain_state->marking_done
+              && domain_state->sweeping_done));
```

Comment:
This should probably also just use the new helper function `marking_and_sweeping_done(domain_state)` to avoid duplicating the condition.

Reply:
Done, thanks!

For the record, I thought of adding a similar assertion after the `if (last)` case below -- because that case supposedly restores the invariant in the `(last)` case that were established in the other case by the main loop. But I don't think it actually holds: what we need to call `caml_free_shared_heap` is that there are no pools left on the current domain (in particular no unswept pools), but this does not necessarily entail that the `->sweeping_done` boolean has been set.

Now that I write this comment, I see that I probably need a different assertion `caml_shared_heap_is_empty(caml_domain_state*)` to clarify the nuance. I'll try to add that.

Reply:
Done, in a new commit.